src/kvstore/listener/elasticsearch/ESListener.cpp

```diff
@@ -25,48 +25,112 @@ void ESListener::init() {
   if (!cRet.ok() || cRet.value().empty()) {
     LOG(FATAL) << "elasticsearch clients error";
   }
+  std::vector<nebula::plugin::ESClient> esClients;
   for (const auto& c : cRet.value()) {
-    nebula::plugin::HttpClient hc;
-    hc.host = c.host;
+    auto host = c.host;
+    std::string user, password;
     if (c.user_ref().has_value()) {
-      hc.user = *c.user_ref();
-      hc.password = *c.pwd_ref();
+      user = *c.user_ref();
+      password = *c.pwd_ref();
     }
-    hc.connType = c.conn_type_ref().has_value() ? *c.get_conn_type() : "http";
-    esClients_.emplace_back(std::move(hc));
+    std::string protocol = c.conn_type_ref().has_value() ? *c.get_conn_type() : "http";
+    esClients.emplace_back(HttpClient::instance(), protocol, host.toString(), user, password);
   }
-
+  esAdapter_.setClients(std::move(esClients));
   auto sRet = schemaMan_->toGraphSpaceName(spaceId_);
   if (!sRet.ok()) {
     LOG(FATAL) << "space name error";
   }
   spaceName_ = std::make_unique<std::string>(sRet.value());
 }
 
-bool ESListener::apply(const std::vector<KV>& data) {
-  std::vector<nebula::plugin::DocItem> docItems;
-  for (const auto& kv : data) {
-    if (!nebula::NebulaKeyUtils::isTag(vIdLen_, kv.first) &&
-        !nebula::NebulaKeyUtils::isEdge(vIdLen_, kv.first)) {
-      continue;
+bool ESListener::apply(const BatchHolder& batch) {
+  nebula::plugin::ESBulk bulk;
+  auto callback = [&bulk](BatchLogType type,
+                          const std::string& index,
+                          const std::string& vid,
+                          const std::string& src,
+                          const std::string& dst,
+                          int64_t rank,
+                          const std::string& text) {
+    if (type == BatchLogType::OP_BATCH_PUT) {
+      bulk.put(index, vid, src, dst, rank, text);
+    } else if (type == BatchLogType::OP_BATCH_REMOVE) {
+      bulk.delete_(index, vid, src, dst, rank);
+    } else {
+      LOG(FATAL) << "Unexpect";
```

Comment:
You'd better don't write the log of `OP_REMOVE_RANGE` into batch in `processLogs`. Only put the LogType you can handle into the batch is enough.

Reply:
done